### PR TITLE
1912-implement-nirb-tools

### LINKED ISSUE #1912: implement nirb tools in feel++
Non intrusive reduced basis methods could be useful to enable model order reduction on complex models/numerical methods

- [x] feelpp/feelpp#1949
- [x] feelpp/feelpp#1950
- [x] feelpp/feelpp#1951
- [x] feelpp/feelpp#1952
- [x] feelpp/phd.thomas#14
- [x] feelpp/feelpp#1954
- [x] #1955
- [x] Use RIC for POD basis generation
- [ ] #2038

### PR DESCRIPTION
**Pull request by thomas-saigre**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [ ] Have you successfully run the Feel++ testsuite with your changes locally?
- [ ] Have you written Doxygen comments in your contribution ?

This is not working in parallel yet

- close #1912 
-----
[skip feelpp skip tests skip toolboxes skip mor]

### CODE REVIEW

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+import feelpp.toolboxes.fluid as fluid
+import feelpp.interpolation as fi
+
+
+import os
+
+from mpi4py import MPI
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: This is not a good idea, the communicators are given by Feel++ you never know how the method is used. 
The communicator should come from the toolbox

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+
+class ToolboxModel():
+
+    def __init__(self, dimension, H, h, toolboxType, cfg_path, model_path, geo_path, order=1) -> None:
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: The communicator must come from the toolbox and be a member of the class

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+    def initModel(self):
+        """Initialize the model
+        """
+        self.model = feelpp.readJson(self.model_path)
+        self.tbFine = self.setToolbox(self.h)
+        self.Xh = feelpp.functionSpace(mesh=self.tbFine.mesh(), order=self.order)
+        self.Dmu = loadParameterSpace(self.model_path)
+        self.Ndofs = self.tbFine.mesh().numGlobalPoints()
+
+        if feelpp.Environment.isMasterRank():
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: Not a good idea it implies that the communicator is COMM_WORLD. 
it should be `self.worldcomm.isMasterRank()`

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+            print(f"[NIRB] Number of nodes on the fine mesh : {self.Ndofs}")
+
+
+    def initCoarseToolbox(self):
+        """Initialize the rectification problem
+        """
+        self.tbCoarse = self.setToolbox(self.H)
+        if feelpp.Environment.isMasterRank():
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: idem

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 5]
```diff
@@ -0,0 +1,802 @@
...
+            Vh_image = image_tb.spaceTemperature()
+            Vh_domain = domain_tb.spaceTemperature()
+        elif self.toolboxType == "fluid":
+            Vh_image = image_tb.spaceVelocity()
+            Vh_domain = domain_tb.spaceVelocity()
+        interpolator = fi.interpolator(domain = Vh_domain, image = Vh_image, range = image_tb.rangeMeshElements())
+        return interpolator
+
+
+class nirbOffline(ToolboxModel):
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py:246]: add documentation

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 6]
```diff
@@ -0,0 +1,802 @@
...
+        self.method = method
+        self.doRectification = doRectification
+
+        self.l2ScalarProductMatrix = None
+        self.h1ScalarProductMatrix = None
+        self.N = 0 # number of modes
+
+        if self.doRectification:
+            super().initCoarseToolbox()
+        if feelpp.Environment.isMasterRank():
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: `self.worldcomm.isMasterRank()`

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 7]
```diff
@@ -0,0 +1,802 @@
...
+        self.coarseSnapShotList = []
+
+        s = self.Dmu.sampling()
+        s.sampling(numberOfInitSnapshots, samplingMode)
+        vector_mu = s.getVector()
+
+        if computeCoarse:
+            assert self.tbCoarse is not None, f"Coarse toolbox needed for computing coarse Snapshot. set doRectification->True"
+            for mu in vector_mu:
+                if feelpp.Environment.isMasterRank():
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: `self.worldcomm.isMasterRank()`

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 8]
```diff
@@ -0,0 +1,802 @@
...
+    def generateReducedBasis(self, tolerance=1.e-6, regulParam=1.e-10):
+        """Generate the reduced basis, and store it in self.reducedBasis
+
+        Args :
+            tolerance(float), optional : the tolerance value for
+            regulParam(float), optional : the regularization parameter for rectification
+        """
+        self.reducedBasis = self.PODReducedBasis(tolerance=tolerance)
+        self.N = self.reducedBasis.size[1]
+        if feelpp.Environment.isMasterRank():
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: see above

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 9]
```diff
@@ -0,0 +1,802 @@
...
+
+        Nsnap = len(self.fineSnapShotList)
+        correlationMatrix = PETSc.Mat().create()
+        correlationMatrix.setSizes([Nsnap,Nsnap])
+        correlationMatrix.setFromOptions()
+        correlationMatrix.setUp()
+
+        for i, snap1 in enumerate(self.fineSnapShotList):
+            for j, snap2 in enumerate(self.fineSnapShotList):
+                    correlationMatrix[i,j] = self.scalarL2(snap1.to_petsc().vec(),snap2.to_petsc().vec())
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: not a good idea, we have a L2 scalar product for these function

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 10]
```diff
@@ -0,0 +1,802 @@
...
+
+        correlationMatrix.assemble()
+        eigenValues, eigenVectors =  TruncatedEigenV(correlationMatrix, tolerance) 
+
+        Nmode = len(eigenVectors)
+        for i in range(Nmode):
+            eigenVectors[i] /= np.sqrt(np.abs(eigenValues[i]))
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: why do you divide by the eigenvalues ?

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 11]
```diff
@@ -0,0 +1,802 @@
...
+                    int_X(u v)
+
+        Args:
+            u (PETSc.Vec): vector
+            v (PETSC.Vec): second vector
+
+        Returns:
+            float: v.T @ ML2 @ u
+        """
+        return v.dot( self.l2ScalarProductMatrix.to_petsc().mat() * u )   # v.T @ A @ u
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: this function is provided and is called `energy()` 
```c++
l2ScalarProductMatrix.energy(v,u)
```

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 12]
```diff
@@ -0,0 +1,802 @@
...
+                    int_X(\nabla u \nabla v)
+
+        Args:
+            u (PETSc.Vec): vector
+            v (PETSC.Vec): second vector
+
+        Returns:
+            float: v.T @ MH1 @ u
+        """
+        return v.dot( self.h1ScalarProductMatrix.to_petsc().mat() * u )   # v.T @ A @ u
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: idem, see above

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py (hunk 1 above)
Comment (prudhomm) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: @alielarif this should not be there !

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py (hunk 3 above)
Comment (prudhomm) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: this has not been adressed

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp [hunk 13]
```diff
@@ -56,9 +56,11 @@ void defFM(py::module &m)
              )
         .def("init",static_cast<void (fm_t::*)(bool)>(&fm_t::init), "initialize the fluid mechanics toolbox",py::arg("buildModelAlgebraicFactory")= true)
         .def("mesh",static_cast<typename fm_t::mesh_ptrtype  (fm_t::*)() const>(&fm_t::mesh), "get the mesh")
+        .def("rangeMeshElements", &fm_t::rangeMeshElements, "get the range of mesh elements" )
         .def("setMesh",static_cast<void (fm_t::*)(typename fm_t::mesh_ptrtype const&)>(&fm_t::setMesh), "set the mesh of the toolbox",py::arg("mesh"))
+        .def("updateParameterValues", &fm_t::updateParameterValues, "update parameter values" )
         // function spaces and elements
-        .def("functionSpaceVelocity",&fm_t::functionSpaceVelocity, "get the velocity function space")
+        .def("spaceVelocity",&fm_t::functionSpaceVelocity, "get the velocity function space")
```
Comment (prudhomm) [inline python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp:66]: @thomas-saigre @alielarif why the change of name ?

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp [hunk 14]
```diff
@@ -74,7 +76,7 @@ void defFM(py::module &m)
             []( std::shared_ptr<fm_t>& self, typename fm_t::element_pressure_ptrtype& p ) {
                 self->fieldPressure() = *p;
             }, "set the pressure field", py::arg("field"))
-        .def("functionSpacePressure",&fm_t::functionSpacePressure, "get the pressure function space")
+        .def("spacePressure",&fm_t::functionSpacePressure, "get the pressure function space")
```
Comment (prudhomm) [inline python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp:82]: idem

File: python/pyfeelpp/tests/test_interpolation.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
+import sys,os
+import feelpp
+import pytest
+from feelpp.toolboxes.core import *
+from feelpp.toolboxes.heat import *
```
Comment (prudhomm) [inline python/pyfeelpp/tests/test_interpolation.py:5]: this test must be in python/pyfeelpp-toolboxes/tests if you use the toolboxes

**Review by prudhomm — CHANGES_REQUESTED** [review]

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py (hunk 3 above)
Comment (alielarif) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: That means we need firstly to define the variable worldcomm by  :
self.worldcomm = feelpp.Environment.worldComm() ?

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py (hunk 4 above)
Comment (alielarif) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: done.

File: toolboxes/fluid/cases/lid-driven-cavity/cfd2d.json [hunk 16]
```diff
@@ -25,15 +25,23 @@
         "d2r_bounded_wall":"min(meshes_fluid_distanceToRange_wall_normalized_min_max,0.2)/0.2:meshes_fluid_distanceToRange_wall_normalized_min_max",
         "mymetric_wall":"hclose_wall+(hfar-hclose_wall)*d2r_bounded_wall:hclose_wall:hfar:d2r_bounded_wall"
     },
+    "CRBParameters":
```
Comment (vincentchabannes) [inline toolboxes/fluid/cases/lid-driven-cavity/cfd2d.json]: this file should not be modified! and no reason to have CRBParameters here

Comment (vincentchabannes) [reply toolboxes/fluid/cases/lid-driven-cavity/cfd2d.json]: please revert all changes in this file

File: toolboxes/fluid/cases/lid-driven-cavity/cfd2d.cfg [hunk 17]
```diff
@@ -4,6 +4,10 @@ directory=toolboxes/fluid/lid-driven-cavity/2d/P2P1G1
 case.dimension=2
 #case.discretization=P1P1G1
 
+[nirb]
```
Comment (vincentchabannes) [inline toolboxes/fluid/cases/lid-driven-cavity/cfd2d.cfg]: this file should not be modified! please revert

**Review by vincentchabannes — CHANGES_REQUESTED** [review]

File: toolboxes/fluid/cases/lid-driven-cavity/cfd2d.cfg (hunk 17 above)
Comment (thomas-saigre) [reply toolboxes/fluid/cases/lid-driven-cavity/cfd2d.cfg]: done in 941376ed5a8dda927b1888a051e6d760103df348

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 18]
```diff
@@ -0,0 +1,1124 @@
...
+
+        self.fineSnapShotList = []
+        self.coarseSnapShotList = []
+
+        if Xi_train is None:
+            s = self.Dmu.sampling()
+            s.sampling(numberOfInitSnapshots, samplingMode)
+            vector_mu = s.getVector()
+        else:
+            vector_mu = choices(Xi_train, k=numberOfInitSnapshots)
```
Comment (alielarif) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: there is an issue in the function choices. It doesn't return what we expect. 
I give a list of 10 params and ask him to chose 100 params. It didn't crash !!

Comment (thomas-saigre) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: It seems that `random.choices` runs the draw with replacement. To avoid it you can use the numpy function :
```python
np.random.choice(list, size=size, replace=False)
```

File: mor/cases/thermal-fin/3d/thermal-fin.cfg [hunk 19]
```diff
@@ -3,20 +3,23 @@ case.dimension=3
 case.discretization=P1
 journal=false
```
Comment (vincentchabannes) [inline mor/cases/thermal-fin/3d/thermal-fin.cfg]: should be removed (journal allow us to get a lot of info)

File: .vscode/settings.json [hunk 20]
```diff
@@ -107,9 +107,15 @@
...
-        "texture": "cpp"
+        "*.py.in": "python",
+        "texture": "cpp",
+        "charconv": "cpp",
+        "numbers": "cpp",
+        "scoped_allocator": "cpp",
+        "semaphore": "cpp"
     },
-    "C_Cpp.configurationWarnings": "Disabled",
+    "jupyter.notebookFileRoot": "${workspaceFolder}/build/python/install/lib/python3/dist-packages",
```
Comment (thomas-saigre) [inline .vscode/settings.json:117]: @prudhomm @vincentchabannes @alielarif I'm not sure this line should stay here : if the preset selected is not `python` there is no reason this path should be correct.
The advantage of this setting it that when we run a notebook in vscode, it automatically find the development packages installed in `build` directory, so there is no need to set by hand `$PYTHONPATH`.
The drawback is that when we launch a kernel, the working directory is no longer the directory where the ipynb file is saved, it is the one set in this line. So it is messy to load the config file (for example in [python/pyfeelpp-mor/feelpp/mor/nirb/test_nirb_offline.ipynb](https://github.com/feelpp/feelpp/blob/e82a120d93c2791083f631431752c95031d6cd60/python/pyfeelpp-mor/feelpp/mor/nirb/test_nirb_offline.ipynb), the path is hard coded to a local path `"/home/saigre/Documents/code/feelpp/python/pyfeelpp-mor/feelpp/mor/nirb/model/square/square.cfg"` which should have not been pushed, my bad !)

I don't know what would be the best way to set the environment easily : automatically set `PYTHONPATH` to the install directory, while the working directory still being the one where the notebook file is saved...

Comment (prudhomm) [reply .vscode/settings.json:117]: could you set this in your user settings and not the workspace setting ?

File: .vscode/settings.json [hunk 21]
```diff
@@ -119,5 +125,7 @@
     
     ],
     "cmake.buildDirectory": "${workspaceFolder}/build-${buildKitVendor}${buildKitVersionMajor}-${variant:cpp}-${variant:toolboxes}-${variant:mor}-${variant:buildType}-${variant:webserver}-${variant:doxygen}-${variant:matplot}",
-    "cmake.parallelJobs": 5
-}
\ No newline at end of file
+    "cmake.parallelJobs": 5,
+    "autoDocstring.docstringFormat": "numpy"
```
Comment (prudhomm) [inline .vscode/settings.json:129]: why numpy?

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 22]
```diff
@@ -0,0 +1,1282 @@
...
+
+        Returns:
+            feelpp_.discr.Element_*: field of the solution
+        """
+        if self.toolboxType == "heat":
+            return toolbox.fieldTemperature()
+        elif self.toolboxType == "fluid":
+            return toolbox.fieldVelocity()
+        else:
+            raise ValueError("Unknown toolbox")
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py:188]: how about using a dict to map the function calls ? that way it extends in a nicer way to other toolboxes

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 23]
```diff
@@ -0,0 +1,1282 @@
...
+        mesh_ = feelpp.mesh(dim=self.dimension, realdim=self.dimension)
+        mesh = feelpp.load(mesh_, mesh_path, hsize)
+
+        # set mesh and model properties
+        if self.toolboxType == "heat":
+            tb = heat.heat(dim=self.dimension, order=self.order)
+        elif self.toolboxType == "fluid":
+            tb = fluid.fluid(dim=self.dimension)
+        else:
+            raise ValueError("Unknown toolbox")
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py:162]: how about create the toolbox outside ToolboxModel ?

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 24]
```diff
@@ -0,0 +1,1282 @@
...
+
+        Returns:
+            OperatorInterpolation: interpolator object
+        """
+        if self.toolboxType == "heat":
+            Vh_image = image_tb.spaceTemperature()
+            Vh_domain = domain_tb.spaceTemperature()
+        elif self.toolboxType == "fluid":
+            Vh_image = image_tb.spaceVelocity()
+            Vh_domain = domain_tb.spaceVelocity()
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py:221]: this is not good I think to have this everywhere, if we have to extend this to other toolboxes

Comment (prudhomm) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py:221]: A facade needs to be created : please follow this: https://www.geeksforgeeks.org/facade-method-python-design-patterns/

Comment (prudhomm) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py:221]: we need to discuss this before doing any major changes to the code

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py [hunk 25]
```diff
@@ -0,0 +1,1282 @@
...
+        nb = 0
+        pas = 1 if (self.N<50) else 5 
+        for i in tqdm(range(1,self.N+1,pas), desc=f"[NIRB] Compute convergence error:", ascii=False, ncols=100):
+            nb = i
+            # Get rectification matrix
+            BH = self.coeffCoarse[:nb, :nb]
+            Bh = self.coeffFine[:nb, :nb]
+            R = np.zeros((nb,nb))
+            for s in range(nb):
+                R[s,:] = (np.linalg.inv(BH.transpose() @ BH + regulParam*np.eye(nb))@BH.transpose()@Bh[:,s])
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: avoid calling `.inv`, rather call `.solve` with multiple right hand sides, @thomas-saigre

File: python/pyfeelpp-mor/feelpp/mor/nirb/test_process.py [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+## Import module 
+from postProcess import * 
+
+#%%
+
+if __name__ == "__main__":
+    
+    import sys 
+
+    dataPath = "/data/home/elarif/feelppdb/nirb/heat/np_1/results/Rect/"
```
Comment (prudhomm) [inline python/pyfeelpp-mor/feelpp/mor/nirb/test_process.py]: this should not be hardcoded, please use feel++ framework and repositories

File: python/pyfeelpp-mor/tests/test_nirb.py [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+
+@pytest.mark.parametrize("dir,cfg,json,rect,greedy", cases_params_nirb, ids=cases_ids_nirb)
+def test_nirb(dir, cfg, json, rect, greedy, init_feelpp):
+    e = init_feelpp
+    casefile = os.path.join(os.path.dirname(__file__), dir, cfg)
+    model_path = os.path.join(os.path.dirname(__file__), dir, json)
+    feelpp.Environment.setConfigFile(casefile)
+
+    run_offline(model_path, rect, greedy)
+    run_online(model_path, rect)
```
Comment (prudhomm) [inline python/pyfeelpp-mor/tests/test_nirb.py:81]: are there any verifications that the results are ok ?

**Review by prudhomm — CHANGES_REQUESTED** [review]

**Comment by prudhomm** [comment]
@alielarif @thomas-saigre minor changes are required.
for example remove all hardcoded path and use feelpp repository handling instead

File: .vscode/settings.json (hunk 21 above)
Comment (thomas-saigre) [reply .vscode/settings.json:129]: I looked at the list available for default docstring format. When the default one is use, the overview of it in vscode is given in the first figure, with numpy, the format is slightly different and the overview is given in the second figure.
I find this is better the second way, but of course we can choose something else !


image 
image

File: .vscode/settings.json [hunk 28]
```diff
@@ -119,5 +125,7 @@
     
     ],
     "cmake.buildDirectory": "${workspaceFolder}/build-${buildKitVendor}${buildKitVersionMajor}-${variant:cpp}-${variant:toolboxes}-${variant:mor}-${variant:buildType}-${variant:webserver}-${variant:doxygen}-${variant:matplot}",
-    "cmake.parallelJobs": 5
-}
\ No newline at end of file
+    "cmake.parallelJobs": 5,
```
Comment (thomas-saigre) [inline .vscode/settings.json:128]: I think I have pussed this by error : I think it was `25` before (too much for my personal laptop), I thought I updated this value only locally, but it seems that it commited anyway, sorry about that.
Shall we change it back ?

File: python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp (hunk 13 above)
Comment (thomas-saigre) [reply python/pyfeelpp-toolboxes/feelpp/toolboxes/fluid/fluid.cpp:66]: That was after a conversation we had on slack : https://feelpp.slack.com/archives/C6A375EMS/p1662299935892779 :

> @prudhomm juste une petite remarque pour le wrapping python : dans la toolbox heat, pour accéder aux fonction de l'espace, la fonction est spaceTemperature(), tandis que pour la toolbox fluid c'est functionSpaceVelocity(). Est-ce que ce ne serait pas plus cohérent de mettre le même template de nom pour toutes les toolbox (en plus ça perd github copilot !)

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py (hunk 25 above)
Comment (thomas-saigre) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py]: Yes the correction was made in another branch (in commit 1991f58a5869faee625b4c19f20f5b5bb9ad7394 on branch 2038-nirb-greedy), we may cherry-pick this commit to put the correction in here

File: python/pyfeelpp-mor/feelpp/mor/nirb/test_process.py (hunk 26 above)
Comment (alielarif) [reply python/pyfeelpp-mor/feelpp/mor/nirb/test_process.py]: we replace this file by a notebook one in the branch [#2041](https://github.com/feelpp/feelpp/issues/2041). This will be easy to manipulate.

File: python/pyfeelpp-mor/tests/test_nirb.py [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+import feelpp
+from feelpp.mor.nirb.nirb import *
+
+# desc : ((toolboxtype, 'model_directory', cfg, json, geo, H, h, dimension, doRectification, doGreedy), 'name-of-the-test')
+casesNirb = [
+        #  (('testcase/nirb/lid-driven-cavity/', 'cfd2d.cfg', 'cfd2d.json', False), 'lid-driven-cavity w/o rect.'),
+        #  (('testcase/nirb/lid-driven-cavity/', 'cfd2d.cfg', 'cfd2d.json', True) , 'lid-driven-cavity rect'),
+         (('testcase/nirb/square', 'square.cfg', 'square.json', False, False), 'square2d w/o rect wogreedy'),
+         (('testcase/nirb/square', 'square.cfg', 'square.json', True, False) , 'square2d rect wogreedy'),
+         (('testcase/nirb/square', 'square.cfg', 'square.json', True, True) , 'square2d rect egreedy'),
```
Comment (alielarif) [inline python/pyfeelpp-mor/tests/test_nirb.py]: the test with Greedy selection dosn't get an error less than 8.10-2. I remove it now by waiting its new version.

File: python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py (hunk 23 above)
Comment (alielarif) [reply python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py:162]: This will lead to a lot of code changes. Better to do it in an other branch.

**Review by prudhomm — APPROVED** [review]
